// skynet-js/dist/cjs/skylink/sia.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.newSkylinkV2 = exports.newEd25519PublicKey = exports.newSpecifier = exports.SiaSkylink = void 0;
const base64_js_1 = require("base64-js");
const crypto_1 = require("../crypto");
const encoding_1 = require("../utils/encoding");
const string_1 = require("../utils/string");
const validation_1 = require("../utils/validation");
// The raw size of the data that gets put into a link.
const RAW_SKYLINK_SIZE = 34;
class SiaSkylink {
    constructor(bitfield, merkleRoot) {
        this.bitfield = bitfield;
        this.merkleRoot = merkleRoot;
        validation_1.validateNumber("bitfield", bitfield, "constructor parameter");
        validation_1.validateUint8ArrayLen("merkleRoot", merkleRoot, "constructor parameter", 32);
    }
    toBytes() {
        const buf = new ArrayBuffer(RAW_SKYLINK_SIZE);
        const view = new DataView(buf);
        view.setUint16(0, this.bitfield, true);
        const uint8Bytes = new Uint8Array(buf);
        uint8Bytes.set(this.merkleRoot, 2);
        return uint8Bytes;
    }
    toString() {
        let base64 = base64_js_1.fromByteArray(this.toBytes());
        // Change to URL encoding.
        base64 = base64.replace(/\+/g, "-").replace(/\//g, "_");
        // Remove padding characters.
        return string_1.trimSuffix(base64, "=");
    }
}
exports.SiaSkylink = SiaSkylink;
const SPECIFIER_LEN = 16;
/**
 * Returns a specifier for given name, a specifier can only be 16 bytes so we
 * panic if the given name is too long.
 *
 * @param name - The name.
 * @returns - The specifier, if valid.
 */
function newSpecifier(name) {
    validateSpecifier(name);
    const specifier = new Uint8Array(SPECIFIER_LEN);
    specifier.set(string_1.stringToUint8ArrayUtf8(name));
    return specifier;
}
exports.newSpecifier = newSpecifier;
const PUBLIC_KEY_SIZE = 32;
class SiaPublicKey {
    constructor(algorithm, key) {
        this.algorithm = algorithm;
        this.key = key;
    }
    marshalSia() {
        const bytes = new Uint8Array(SPECIFIER_LEN + 8 + PUBLIC_KEY_SIZE);
        bytes.set(this.algorithm);
        bytes.set(encoding_1.encodePrefixedBytes(this.key), SPECIFIER_LEN);
        return bytes;
    }
}
/**
 * Creates a new sia public key. Matches Ed25519PublicKey in sia.
 *
 * @param publicKey - The hex-encoded public key.
 * @returns - The SiaPublicKey.
 */
function newEd25519PublicKey(publicKey) {
    validation_1.validateHexString("publicKey", publicKey, "parameter");
    const algorithm = newSpecifier("ed25519");
    const publicKeyBytes = string_1.hexToUint8Array(publicKey);
    validation_1.validateUint8ArrayLen("publicKeyBytes", publicKeyBytes, "converted publicKey", PUBLIC_KEY_SIZE);
    return new SiaPublicKey(algorithm, publicKeyBytes);
}
exports.newEd25519PublicKey = newEd25519PublicKey;
/**
 * Creates a new v2 skylink. Matches NewSkylinkV2 in skyd.
 *
 * @param siaPublicKey - The public key as a SiaPublicKey.
 * @param tweak - The hashed tweak.
 * @returns - The v2 skylink.
 */
function newSkylinkV2(siaPublicKey, tweak) {
    const version = 2;
    const bitfield = version - 1;
    const merkleRoot = deriveRegistryEntryID(siaPublicKey, tweak);
    return new SiaSkylink(bitfield, merkleRoot);
}
exports.newSkylinkV2 = newSkylinkV2;
/**
 * A helper to derive an entry id for a registry key value pair. Matches `DeriveRegistryEntryID` in sia.
 *
 * @param pubKey - The sia public key.
 * @param tweak - The tweak.
 * @returns - The entry ID as a hash of the inputs.
 */
function deriveRegistryEntryID(pubKey, tweak) {
    return crypto_1.hashAll(pubKey.marshalSia(), tweak);
}
/**
 * Performs validation checks on the specifier name, it panics when the input is
 * invalid seeing we want to catch this on runtime.
 * Matches `validateSpecifier` in sia.
 *
 * @param name - The specifier name.
 * @throws - Will throw if the specifier name is not valid.
 */
function validateSpecifier(name) {
    if (!string_1.isASCIIString(name)) {
        throw new Error("specifier has to be ASCII");
    }
    if (name.length > SPECIFIER_LEN) {
        throw new Error("specifier max length exceeded");
    }
}
