// skynet-js/dist/mjs/mysky/index.js
export { DacLibrary } from "./dac";
import { ParentHandshake, WindowMessenger } from "post-me";
import { dispatchedErrorEvent, errorWindowClosed, monitorWindowError, PermCategory, Permission, PermType, } from "skynet-mysky-utils";
import { Connector, defaultConnectorOptions } from "./connector";
import { defaultGetEntryOptions, defaultSetEntryOptions } from "../registry";
import { defaultGetJSONOptions, defaultSetJSONOptions, getOrCreateRegistryEntry, } from "../skydb";
import { hexToUint8Array } from "../utils/string";
import { deriveDiscoverableTweak } from "./tweak";
import { popupCenter } from "./utils";
import { validateObject, validateOptionalObject, validateString } from "../utils/validation";
import { extractOptions } from "../utils/options";
export const mySkyDomain = "skynet-mysky.hns";
export const mySkyDevDomain = "skynet-mysky-dev.hns";
export const mySkyAlphaDomain = "sandbridge.hns";
const mySkyUiRelativeUrl = "ui.html";
const mySkyUiTitle = "MySky UI";
const [mySkyUiW, mySkyUiH] = [600, 600];
export async function loadMySky(skappDomain, customOptions) {
    const mySky = await MySky.New(this, skappDomain, customOptions);
    return mySky;
}
export class MySky {
    // ============
    // Constructors
    // ============
    constructor(connector, permissions, hostDomain) {
        this.connector = connector;
        this.hostDomain = hostDomain;
        this.dacs = [];
        this.grantedPermissions = [];
        this.pendingPermissions = permissions;
    }
    static async New(client, skappDomain, customOptions) {
        const opts = { ...defaultConnectorOptions, ...customOptions };
        // Enforce singleton.
        if (MySky.instance) {
            return MySky.instance;
        }
        let domain = mySkyDomain;
        if (opts.alpha) {
            domain = mySkyAlphaDomain;
        }
        else if (opts.dev) {
            domain = mySkyDevDomain;
        }
        const connector = await Connector.init(client, domain, customOptions);
        const hostDomain = await client.extractDomain(window.location.hostname);
        const permissions = [];
        if (skappDomain) {
            // TODO: Are these permissions correct?
            const perm1 = new Permission(hostDomain, skappDomain, PermCategory.Hidden, PermType.Read);
            const perm2 = new Permission(hostDomain, skappDomain, PermCategory.Hidden, PermType.Write);
            permissions.push(perm1, perm2);
        }
        MySky.instance = new MySky(connector, permissions, hostDomain);
        return MySky.instance;
    }
    // ==========
    // Public API
    // ==========
    /**
     * Loads the given DACs.
     */
    async loadDacs(...dacs) {
        const promises = [];
        for (const dac of dacs) {
            promises.push(this.loadDac(dac));
        }
        this.dacs.push(...dacs);
        await Promise.all(promises);
    }
    async addPermissions(...permissions) {
        this.pendingPermissions.push(...permissions);
    }
    async checkLogin() {
        const [seedFound, permissionsResponse] = await this.connector.connection.remoteHandle().call("checkLogin", this.pendingPermissions);
        // Save granted and failed permissions.
        const { grantedPermissions, failedPermissions } = permissionsResponse;
        this.grantedPermissions = grantedPermissions;
        this.pendingPermissions = failedPermissions;
        const loggedIn = seedFound && failedPermissions.length === 0;
        this.handleLogin(loggedIn);
        return loggedIn;
    }
    /**
     * Destroys the mysky connection by:
     *
     * 1. Destroying the connected DACs,
     *
     * 2. Closing the connection,
     *
     * 3. Closing the child iframe
     */
    async destroy() {
        // TODO: For all connected dacs, send a destroy call.
        // TODO: Delete all connected dacs.
        // Close the connection.
        this.connector.connection.close();
        // Close the child iframe.
        if (this.connector.childFrame) {
            this.connector.childFrame.parentNode.removeChild(this.connector.childFrame);
        }
    }
    async logout() {
        return await this.connector.connection.remoteHandle().call("logout");
    }
    async requestLoginAccess() {
        let uiWindow;
        let uiConnection;
        let seedFound = false;
        // Add error listener.
        const { promise: promiseError, controller: controllerError } = monitorWindowError();
        // eslint-disable-next-line no-async-promise-executor
        const promise = new Promise(async (resolve, reject) => {
            // Make this promise run in the background and reject on window close or any errors.
            promiseError.catch((err) => {
                if (err === errorWindowClosed) {
                    // Resolve without updating the pending permissions.
                    resolve();
                    return;
                }
                reject(err);
            });
            try {
                // Launch the UI.
                uiWindow = await this.launchUI();
                uiConnection = await this.connectUi(uiWindow);
                // Send the UI the list of required permissions.
                // TODO: This should be a dual-promise that also calls ping() on an interval and rejects if no response was found in a given amount of time.
                const [seedFoundResponse, permissionsResponse] = await uiConnection.remoteHandle().call("requestLoginAccess", this.pendingPermissions);
                seedFound = seedFoundResponse;
                // Save failed permissions.
                const { grantedPermissions, failedPermissions } = permissionsResponse;
                this.grantedPermissions = grantedPermissions;
                this.pendingPermissions = failedPermissions;
                resolve();
            }
            catch (err) {
                reject(err);
            }
        });
        await promise
            .catch((err) => {
            throw err;
        })
            .finally(() => {
            // Close the window.
            if (uiWindow) {
                uiWindow.close();
            }
            // Close the connection.
            if (uiConnection) {
                uiConnection.close();
            }
            // Clean up the event listeners and promises.
            controllerError.cleanup();
        });
        const loggedIn = seedFound && this.pendingPermissions.length === 0;
        this.handleLogin(loggedIn);
        return loggedIn;
    }
    async userID(opts) {
        return await this.connector.connection.remoteHandle().call("userID", opts);
    }
    /**
     * Gets Discoverable JSON at the given path through MySky, if the user has given permissions to do so.
     *
     * @param path - The data path.
     * @param [customOptions] - Additional settings that can optionally be set.
     * @returns - An object containing the json data as well as the skylink for the data.
     */
    async getJSON(path, customOptions) {
        validateString("path", path, "parameter");
        validateOptionalObject("customOptions", customOptions, "parameter", defaultGetJSONOptions);
        const opts = {
            ...defaultGetJSONOptions,
            ...this.connector.client.customOptions,
            ...customOptions,
        };
        const publicKey = await this.userID();
        const dataKey = deriveDiscoverableTweak(path);
        opts.hashedDataKeyHex = true; // Do not hash the tweak anymore.
        return await this.connector.client.db.getJSON(publicKey, dataKey, opts);
    }
    /**
     * Gets the entry link for the entry at the given path. This is a v2 skylink.
     * This link stays the same even if the content at the entry changes.
     *
     * @param path - The data path.
     * @returns - The entry link.
     */
    async getEntryLink(path) {
        validateString("path", path, "parameter");
        const publicKey = await this.userID();
        const dataKey = deriveDiscoverableTweak(path);
        const opts = defaultGetEntryOptions;
        opts.hashedDataKeyHex = true; // Do not hash the tweak anymore.
        return await this.connector.client.registry.getEntryLink(publicKey, dataKey, opts);
    }
    /**
     * Sets Discoverable JSON at the given path through MySky, if the user has given permissions to do so.
     *
     * @param path - The data path.
     * @param json - The json to set.
     * @param [customOptions] - Additional settings that can optionally be set.
     * @returns - An object containing the json data as well as the skylink for the data.
     */
    async setJSON(path, json, customOptions) {
        validateString("path", path, "parameter");
        validateObject("json", json, "parameter");
        validateOptionalObject("customOptions", customOptions, "parameter", defaultSetJSONOptions);
        const opts = {
            ...defaultSetJSONOptions,
            ...this.connector.client.customOptions,
            ...customOptions,
        };
        const publicKey = await this.userID();
        const dataKey = deriveDiscoverableTweak(path);
        opts.hashedDataKeyHex = true; // Do not hash the tweak anymore.
        const [entry, dataLink] = await getOrCreateRegistryEntry(this.connector.client, hexToUint8Array(publicKey), dataKey, json, opts);
        const signature = await this.signRegistryEntry(entry, path);
        const setEntryOpts = extractOptions(opts, defaultSetEntryOptions);
        await this.connector.client.registry.postSignedEntry(publicKey, entry, signature, setEntryOpts);
        return { data: json, dataLink };
    }
    // ================
    // Internal Methods
    // ================
    async catchError(errorMsg) {
        const event = new CustomEvent(dispatchedErrorEvent, { detail: errorMsg });
        window.dispatchEvent(event);
    }
    async launchUI() {
        const mySkyUrl = new URL(this.connector.url);
        mySkyUrl.pathname = mySkyUiRelativeUrl;
        const uiUrl = mySkyUrl.toString();
        // Open the window.
        const childWindow = popupCenter(uiUrl, mySkyUiTitle, mySkyUiW, mySkyUiH);
        if (!childWindow) {
            throw new Error(`Could not open window at '${uiUrl}'`);
        }
        return childWindow;
    }
    async connectUi(childWindow) {
        const options = this.connector.options;
        // Complete handshake with UI window.
        const messenger = new WindowMessenger({
            localWindow: window,
            remoteWindow: childWindow,
            remoteOrigin: "*",
        });
        const methods = {
            catchError: this.catchError,
        };
        const connection = await ParentHandshake(messenger, methods, options.handshakeMaxAttempts, options.handshakeAttemptsInterval);
        return connection;
    }
    async loadDac(dac) {
        // Initialize DAC.
        await dac.init(this.connector.client, this.connector.options);
        // Add DAC permissions.
        const perms = dac.getPermissions();
        this.addPermissions(...perms);
    }
    handleLogin(loggedIn) {
        if (loggedIn) {
            for (const dac of this.dacs) {
                dac.onUserLogin();
            }
        }
    }
    async signRegistryEntry(entry, path) {
        return await this.connector.connection.remoteHandle().call("signRegistryEntry", entry, path);
    }
}
MySky.instance = null;
