// skynet-js/dist/cjs/mysky/tweak.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.deriveDiscoverableTweak = exports.hashPathComponent = exports.splitPath = exports.DiscoverableBucketTweak = void 0;
const crypto_1 = require("../crypto");
const string_1 = require("../utils/string");
const discoverableBucketTweakVersion = 1;
class DiscoverableBucketTweak {
    constructor(path) {
        const paths = splitPath(path);
        const pathHashes = paths.map(hashPathComponent);
        this.version = discoverableBucketTweakVersion;
        this.path = pathHashes;
    }
    encode() {
        const size = 1 + 32 * this.path.length;
        const buf = new Uint8Array(size);
        buf.set([this.version]);
        let offset = 1;
        for (const pathLevel of this.path) {
            buf.set(pathLevel, offset);
            offset += 32;
        }
        return buf;
    }
    getHash() {
        const encoding = this.encode();
        return crypto_1.hashAll(encoding);
    }
}
exports.DiscoverableBucketTweak = DiscoverableBucketTweak;
function splitPath(path) {
    return path.split("/");
}
exports.splitPath = splitPath;
function hashPathComponent(component) {
    return crypto_1.hashAll(string_1.stringToUint8ArrayUtf8(component));
}
exports.hashPathComponent = hashPathComponent;
function deriveDiscoverableTweak(path) {
    const dbt = new DiscoverableBucketTweak(path);
    const bytes = dbt.getHash();
    return string_1.toHexString(bytes);
}
exports.deriveDiscoverableTweak = deriveDiscoverableTweak;
